cub/cub/device/device_scan.cuh

```diff
@@ -1164,6 +1164,110 @@ struct DeviceScan
+  //! is assigned to ``*d_dout``.
+  //!
+  //! - Supports non-commutative scan operators.
+  //! - Results are not deterministic for pseudo-associative operators (e.g.,
+  //!   addition of floating-point types). Results for pseudo-associative
+  //!   operators may vary from run to run. Additional details can be found in
+  //!   the @lookback description.
+  //! - When ``d_in`` and ``d_out`` are equal, the scan is performed in-place. The
+  //!   range ``[d_in, d_in + num_items)`` and ``[d_out, d_out + num_items)``
+  //!   shall not overlap in any other way.
+  //! - Note that because initial value is the last argument the cudaStream_t
+  //!   argument also needs be passed explicitly.
+  //! - @devicestorage
+  //!
+  //! Snippet
+  //! +++++++++++++++++++++++++++++++++++++++++++++
+  //!
+  //! The code snippet below illustrates the inclusive max-scan of an ``int`` device vector.
+  //!
+  //! .. literalinclude:: ../../../cub/test/catch2_test_device_scan_api.cu
+  //!     :language: c++
+  //!     :dedent:
+  //!     :start-after: example-begin device-inclusive-scan
+  //!     :end-before: example-end device-inclusive-scan
+  //!
+  //! @endrst
+  //!
+  //! @tparam InputIteratorT
+  //!   **[inferred]** Random-access input iterator type for reading scan inputs @iterator
+  //!
+  //! @tparam OutputIteratorT
+  //!   **[inferred]** Random-access output iterator type for writing scan outputs @iterator
+  //!
+  //! @tparam ScanOpT
+  //!   **[inferred]** Binary scan functor type having member `T operator()(const T &a, const T &b)`
+  //!
+  //! @tparam InitValueT
+  //!  **[inferred]** Type of the `init_value` used Binary scan functor type
+  //!   having member `T operator()(const T &a, const T &b)`
+  //!
+  //! @param[in] d_temp_storage
+  //!   Device-accessible allocation of temporary storage.
+  //!   When `nullptr`, the required allocation size is written to
+  //!   `temp_storage_bytes` and no work is done.
+  //!
+  //! @param[in,out] temp_storage_bytes
+  //!   Reference to size in bytes of `d_temp_storage` allocation
+  //!
+  //! @param[in] d_in
+  //!   Random-access iterator to the input sequence of data items
+  //!
+  //! @param[out] d_out
+  //!   Random-access iterator to the output sequence of data items
+  //!
+  //! @param[in] scan_op
+  //!   Binary scan functor
+  //!
+  //! @param[in] num_items
+  //!   Total number of input items (i.e., the length of `d_in`)
+  //!
+  //! @param[in] stream
+  //!   CUDA stream to launch kernels within.
+  //!
+  //! @param[in] init_value
+  //!   Initial value to seed the inclusive scan (and is assigned to `*d_out`)
+  template <typename InputIteratorT, typename OutputIteratorT, typename ScanOpT, typename InitValueT>
+  CUB_RUNTIME_FUNCTION static cudaError_t InclusiveScan(
+    void* d_temp_storage,
+    size_t& temp_storage_bytes,
+    InputIteratorT d_in,
+    OutputIteratorT d_out,
+    ScanOpT scan_op,
+    int num_items,
+    cudaStream_t stream,
+    InitValueT init_value)
```

Comment:
critical: we have a debug-synchronous overload expecting boolean parameter after stream. If user passed 0 instead of bool, we'd select a different overload, potentially changing a behavior of the algorithm: 

```cpp
// 0 used to mean debug synchronous but now it's initial value
cub::DeviceScan::InclusiveScan(d_temp_storage, temp_storage_bytes, d_data, d_data, max{}, num_items, stream, 0 /* was debug_synchronous but now init_value*/);
```

Opposite problem happens when initial value is bool. In this case we invoke deprecated API instead of the new API:

```cpp
bool init_value = 0;
cub::DeviceScan::InclusiveScan(d_temp_storage, temp_storage_bytes, d_data, d_data, cub::Sum{}, num_items, stream, init_value /* but treated as debug_synchronous */);
```

We should find another way